src/lib/CoreInstance.cc

```diff
@@ -0,0 +1,304 @@
+    // Iterate over each reservation station in config
+    auto reservation_station = config_["Reservation-Stations"][i];
+    for (size_t j = 0; j < reservation_station["Ports"].size(); j++) {
+      // Iterate over issue ports in reservation station
+      uint8_t port = reservation_station["Ports"][j].as<uint8_t>();
+      if (rsArrangement.size() < port + 1) {
+        // Resize vector to match number of execution ports available across
+        // all reservation stations
+        rsArrangement.resize(port + 1);
+      }
+      // Map an execution port to a reservation station
+      rsArrangement[port] = {i, reservation_station["Size"].as<uint16_t>()};
+    }
+  }
+
+  // Construct the core object based on the defined simulation mode
+  uint64_t entryPoint = process_->getEntryPoint();
+  if (mode_ == SimulationMode::Emulation) {
+    core_ = std::make_shared<simeng::models::emulation::Core>(
+        *instructionMemory_, *dataMemory_, entryPoint, processMemorySize_,
+        *arch_);
+  } else if (mode_ == SimulationMode::InOrderPipelined) {
+    core_ = std::make_shared<simeng::models::inorder::Core>(
+        *instructionMemory_, *dataMemory_, processMemorySize_, entryPoint,
+        *arch_, *predictor_);
+  } else if (mode_ == SimulationMode::OutOfOrder) {
+    core_ = std::make_shared<simeng::models::outoforder::Core>(
+        *instructionMemory_, *dataMemory_, processMemorySize_, entryPoint,
+        *arch_, *predictor_, *portAllocator_, rsArrangement, config_);
+  }
+
+  createSpecialFileDirectory();
+
+  return;
+}
+
+void CoreInstance::createSpecialFileDirectory() {
+  simeng::SpecialFileDirGen SFdir = simeng::SpecialFileDirGen(config_);
+  // Create the Special Files directory if indicated to do so in Config
+  if (config_["CPU-Info"]["Generate-Special-Dir"].as<bool>() == true) {
+    // Remove any current special files dir
+    SFdir.RemoveExistingSFDir();
+    // Create new special files dir
+    SFdir.GenerateSFDir();
+  }
+
+  return;
+}
+
+const SimulationMode CoreInstance::getSimulationMode() const { return mode_; }
+
+std::shared_ptr<simeng::Core> CoreInstance::getCore() const {
+  if (manualCreateCore_ && core_ == nullptr) {
+    std::cerr << "Core object not constructed and marked as needed to be "
+                 "manually created via the createCore() function. If either "
+                 "data or instruction memory interfaces are marked as an "
+                 "`External` type, they must be set manually and then core's "
+                 "creation must be called manually."
+              << std::endl;
+    exit(1);
+  }
+  return core_;
+}
+
+std::shared_ptr<simeng::MemoryInterface> CoreInstance::getDataMemory() const {
+  if (setDataMemory_) {
+    std::cerr << "`External` data memory object not set." << std::endl;
+    exit(1);
+  }
+  return dataMemory_;
+}
+
+std::shared_ptr<simeng::MemoryInterface> CoreInstance::getInstructionMemory()
+    const {
+  if (setInstructionMemory_) {
```

Comment:
It seems this clause will always trigger if external instructionMemory is to be used? Which I presume is incorrect?